include/particles/twophase_particle.tcc

```diff
@@ -0,0 +1,726 @@
+  bool status = false;
+  try {
+    // Assign velocity
+    liquid_velocity_ = velocity;
+    status = true;
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+// Compute mass of particle (both solid and fluid)
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::compute_mass() noexcept {
+  // Check if particle volume is set and material ptr is valid
+  assert(volume_ != std::numeric_limits<double>::max() &&
+         this->material() != nullptr &&
+         this->material(mpm::ParticlePhase::Liquid) != nullptr);
+  // Mass = volume of particle * mass_density
+  // Solid mass
+  this->mass_density_ =
+      (this->material())->template property<double>(std::string("density")) *
+      (1 - this->porosity_);
+  this->mass_ = volume_ * mass_density_;
+
+  // Liquid mass
+  this->liquid_mass_density_ =
+      liquid_saturation_ * porosity_ *
+      (this->material(mpm::ParticlePhase::Liquid))
+          ->template property<double>(std::string("density"));
+  this->liquid_mass_ = volume_ * liquid_mass_density_;
+}
+
+//! Map particle mass and momentum to nodes
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_mass_momentum_to_nodes() noexcept {
+  mpm::Particle<Tdim>::map_mass_momentum_to_nodes();
+  this->map_liquid_mass_momentum_to_nodes();
+}
+
+//! Map liquid mass and momentum to nodes
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_liquid_mass_momentum_to_nodes() noexcept {
+  // Check if liquid mass is set and positive
+  assert(liquid_mass_ != std::numeric_limits<double>::max());
+
+  // Map liquid mass and momentum to nodes
+  for (unsigned i = 0; i < nodes_.size(); ++i) {
+    nodes_[i]->update_mass(true, mpm::ParticlePhase::Liquid,
+                           liquid_mass_ * shapefn_[i]);
+    nodes_[i]->update_momentum(true, mpm::ParticlePhase::Liquid,
+                               liquid_mass_ * shapefn_[i] * liquid_velocity_);
+  }
+}
+
+//! Compute pore pressure
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::compute_pore_pressure(double dt) noexcept {
+  // Check if liquid material and cell pointer are set and positive
+  assert(this->material(mpm::ParticlePhase::Liquid) != nullptr &&
+         cell_ != nullptr);
+
+  // get the bulk modulus of liquid
+  double K = this->material(mpm::ParticlePhase::Liquid)
+                 ->template property<double>(std::string("bulk_modulus"));
+
+  // Compute at centroid
+  // get liquid phase strain rate at cell centre
+  auto liquid_strain_rate_centroid =
+      this->compute_strain_rate(dn_dx_centroid_, mpm::ParticlePhase::Liquid);
+
+  // update pressure
+  this->state_variables_[mpm::ParticlePhase::Liquid].at("pressure") +=
+      -dt * (K / porosity_) *
+      ((1 - porosity_) * strain_rate_.head(Tdim).sum() +
+       porosity_ * liquid_strain_rate_centroid.head(Tdim).sum());
```

Comment:
I think we should separate the `liquid_strain` computation as this will also be used by the semi_implicit solver anyway right? Maybe combine it in a derived `compute_strain` function using a private function, just like `compute_mass` and `compute_liquid_mass`?

Reply:
This is resolved after further discussions